indico/modules/events/papers/client/js/components/CommentForm.jsx

```diff
@@ -0,0 +1,148 @@
+  const canJudge = useSelector(canJudgePaper);
+  const InputComponent = commentFormVisible ? FinalTextArea : FinalInput;
+  const inputProps = commentFormVisible
+    ? {autoFocus: true}
+    : {
+        onFocus: () => {
+          setCommentFormVisible(true);
+          onToggleExpand(true);
+        },
+      };
+
+  const handleSubmit = async (formData, form) => {
+    const error = await onSubmit(formData, form);
+    if (error) {
+      return error;
+    }
+
+    setCommentFormVisible(false);
+    if (onToggleExpand) {
+      onToggleExpand(false);
+    }
+
+    setTimeout(() => form.reset(), 0);
+  };
+
+  const initialValues = comment ? {comment: comment.text} : {};
+  if (!canJudge) {
+    delete visibilityOptions.judges;
+  }
+
+  // you can't change the comment's visiblity if you can't review the paper,
+  // thus there is no point in setting the initial value for the "visibility" field
+  if (canReview) {
+    initialValues.visibility =
+      comment && comment.visibility ? comment.visibility.name : Object.keys(visibilityOptions)[0];
+  }
+
+  return (
+    <div className="f-self-stretch">
+      <FinalForm
+        onSubmit={handleSubmit}
+        initialValues={initialValues}
+        subscription={{submitting: true}}
+      >
+        {fprops => (
+          <Form onSubmit={fprops.handleSubmit}>
+            <InputComponent
+              {...inputProps}
+              name="comment"
+              placeholder={Translate.string('Leave a comment...')}
+              required
+            />
+            {commentFormVisible && (
+              <>
+                {canReview && (
+                  <FinalDropdown
+                    name="visibility"
+                    options={Object.values(visibilityOptions)}
+                    selection
+                    required
+                  />
+                )}
+                <Form.Group style={{marginBottom: 0}} inline>
+                  <FinalSubmitButton
+                    label={
+                      comment ? Translate.string('Update comment') : Translate.string('Comment')
+                    }
+                  />
+                  <Button
+                    disabled={fprops.submitting}
+                    content={Translate.string('Cancel')}
+                    onClick={() => {
+                      setCommentFormVisible(false);
+                      if (onToggleExpand) {
+                        onToggleExpand(false);
+                      }
+                      fprops.form.reset();
+                    }}
+                  />
+                </Form.Group>
+              </>
+            )}
+          </Form>
+        )}
+      </FinalForm>
+    </div>
+  );
+}
+
+CommentForm.propTypes = {
+  onSubmit: PropTypes.func.isRequired,
+  onToggleExpand: PropTypes.func,
+  comment: PropTypes.shape({
+    text: PropTypes.string,
+    visibility: PropTypes.shape({
+      name: PropTypes.string.isRequired,
+    }),
```

Comment:
Just curious, why is this nested into an object instead of just the visibility.

Reply:
I am using other visibility properties like `title` somewhere else. Here I only need the `name`

Reply:
OK, didn't realize there's more than the name sent back from the server.